### ISSUE TITLE
predict() NMS ~5× slower on 8.4.81+ — fast torchvision.ops.nms path is gated on "torchvision" in sys.modules, which lazy imports no longer satisfy

### ISSUE THREAD

**Issue by PavanDendi** [ultralytics/ultralytics#25022]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Between ultralytics **8.4.80 and 8.4.84**, `predict()` NMS became **~5× slower** for high-candidate-count workloads, with **no change to the NMS code itself**. The NMS dispatch selects the fast CUDA `torchvision.ops.nms` **only if `"torchvision" in sys.modules`**, otherwise a pure-torch fallback (`TorchNMS`, added ~8.3.190). In 8.4.81+ the torchvision imports were made lazy/scoped (the same laziness that fixed SAM/optional-import startup), so during a plain `YOLO(...).predict()` **torchvision is not in `sys.modules`** → the gate silently picks the slow pure-torch path.

**Measurements** — same TensorRT `.engine`, same container/hardware, only the ultralytics wheel swapped (Jetson Orin Nano, JetPack 7.2; the mechanism is hardware-independent):

| ultralytics | NMS stage | throughput |
|---|---|---|
| 8.4.80 | ~4 ms | 66.6 FPS |
| 8.4.84 | ~22 ms | 30.7 FPS |
| 8.4.84 **+ `import torchvision`** | ~4 ms | 65.5 FPS |

Workload: `yolov8n` @ `imgsz=640` fed a low confidence floor (`conf≈0.1` → hundreds of candidate boxes) — exactly where the pure-torch NMS is worst.

**Proof it's the `sys.modules` gate:** forcing `import torchvision` (caller top / `sitecustomize`) on 8.4.84 drops the NMS stage **22→4 ms** and restores **65.5 FPS**, matching 8.4.80. No other change.

### Environment

ultralytics 8.4.84 (and any 8.4.81+); torch 2.x + torchvision (CUDA); Python 3.12; reproduces on any platform (the fallback is pure-torch, not Jetson-specific).

### Minimal Reproducible Example

```python
import time, torch
from ultralytics import YOLO        # do NOT import torchvision here
m = YOLO("yolov8n.pt")
im = "https://ultralytics.com/images/bus.jpg"
for _ in range(30):
    t = time.time()
    m.predict(im, conf=0.05, imgsz=640, verbose=False)  # low conf -> many candidates
    torch.cuda.synchronize()
    print(f"{(time.time()-t)*1000:.1f} ms")
# Add `import torchvision` at the top and re-run: NMS/predict is ~5x faster.
```

### Additional

**Suggested fix**

Don't gate the fast path on `"torchvision" in sys.modules`. Attempt `import torchvision` (try/except) at the point NMS is dispatched — or import it in the NMS module — so the fast CUDA path is used whenever torchvision is installed, regardless of whether some other code path imported it first. The `sys.modules` check is fragile to precisely the lazy-import refactors being done.

**Related:** #19747 (torchvision must be imported before a TorchScript-NMS model loads — same gate, different symptom) and the SAM lazy-import change (which surfaced this). #5059 / #18924 / #22069 are a *separate* `torchvision::nms` CUDA-backend failure, not this slow-path selection.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @PavanDendi, thank you for your detailed bug report and for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

Thank you for including a clear description, measurements, environment details, and a minimal reproducible example for the `predict()` NMS slowdown. Since this is a 🐛 Bug Report, please ensure the provided [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) is complete and runnable in a fresh environment so we can debug it efficiently.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

This is an automated response, and an Ultralytics engineer will also assist soon 👀.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Hello.

Thanks for debugging and reporting.

You can try the fix here: https://github.com/ultralytics/ultralytics/pull/25023

```
pip install git+https://github.com/ultralytics/ultralytics@nms-predict-fix
```

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in [PR #25023](https://github.com/ultralytics/ultralytics/pull/25023), authored by @Y-T-G with contributions from @glenn-jocher.

Key changes that may resolve this issue:

- `torchvision` is now conditionally preloaded during `AutoBackend.warmup()` for models that are **not end-to-end**.
- This helps ensure `torchvision` NMS is already available after warmup, reducing first-inference latency for standard NMS paths.
- End-to-end models remain unaffected, since they do not use the same post-processing NMS path.
- NMS comments were updated to clarify the intended behavior:
  - use `torchvision` NMS when already imported,
  - fall back to `TorchNMS` when `torchvision` has not been loaded, avoiding unnecessary import overhead during prediction.

To test whether this resolves the issue, you can install the latest `main` branch:

```bash
pip install git+https://github.com/ultralytics/ultralytics.git@main
```

Alternatively, you can wait for the next official PyPI release and upgrade with:

```bash
pip install -U ultralytics
```

Please let us know whether the changes in this PR resolve the issue in your environment. Thanks 🙏 for reporting this, and any further feedback is welcome if the issue persists.
